server/src/main/scala/org/ensime/server/WebSocketFrameHandler.scala

```diff
@@ -0,0 +1,108 @@
+// Copyright: 2010 - 2016 https://github.com/ensime/ensime-server/graphs
+// License: http://www.gnu.org/licenses/gpl-3.0.en.html
+package org.ensime.server
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+import io.netty.channel.{ Channel, ChannelHandlerContext, SimpleChannelInboundHandler }
+import io.netty.handler.codec.http.websocketx.WebSocketServerProtocolHandler.ServerHandshakeStateEvent.HANDSHAKE_COMPLETE
+import io.netty.handler.codec.http.websocketx.{ WebSocketServerHandshaker, TextWebSocketFrame, WebSocketFrame }
+import io.netty.util.AttributeKey
+
+import org.ensime.api._
+import org.ensime.server.WebServer._
+
+class WebSocketFrameHandler(
+    hookHandlers: HookHandlers
+) extends SimpleChannelInboundHandler[WebSocketFrame] {
+
+  val handshakerKey: AttributeKey[WebSocketServerHandshaker] =
+    AttributeKey.valueOf(classOf[WebSocketServerHandshaker], "HANDSHAKER");
+
+  val inHandlerKey: AttributeKey[String => Unit] =
+    AttributeKey.valueOf(classOf[String => Unit], "INHANDLER");
+
+  private def getSelectedSubprotocol(ctx: ChannelHandlerContext): String =
+    ctx.channel().attr(handshakerKey).get().selectedSubprotocol()
+
+  private def setInHandler(ctx: ChannelHandlerContext, inHandler: String => Unit): Unit =
+    ctx.channel().attr(inHandlerKey).set(inHandler)
+
+  private def getInHandler(ctx: ChannelHandlerContext): String => Unit =
+    ctx.channel().attr(inHandlerKey).get()
+
+  private def encoderFor(subprotocol: String): SubprotocolEncoder =
+    subprotocol match {
+      case "jerky" => JerkySubprotocolEncoder
+      case "swanky" => SwankySubprotocolEncoder
+    }
+
+  private def encodedOutHandler(ch: Channel, encoder: SubprotocolEncoder): OutgoingHandler = {
+    rpcResp =>
+      val response = encoder.writeFrame(rpcResp)
+      ch.writeAndFlush(new TextWebSocketFrame(response))
+  }
+
+  private def encodedInHandler(inHandler: IncomingHandler, encoder: SubprotocolEncoder): String => Unit = {
+    frameText =>
+      val rpcReq = encoder.readFrame(frameText)
+      inHandler(rpcReq)
+  }
+
+  override protected def userEventTriggered(ctx: ChannelHandlerContext, evt: Object): Unit = {
+    if (HANDSHAKE_COMPLETE == evt) {
+      Future {
+        // Sorry about this tricky hack, it's the only way I found to access the handshaker
+        // associated with the channel so I can ask it for the selected subprotocol.
+        // The thing is that for the time when we receive this event the handshaker wasn't
+        // set as a channel attribute yet. So we MUST wait for it for a few millis, and for
+        // some reason we MUST do it in a separate thread.
+        // See: [WebSocketServerProtocolHandshakeHandler.java:88](https://git.io/v6DcJ)
```

Comment:
I've opened an issue for it (netty/netty#5734), it shouldn't be that complicated. Anyways, it's not like they answer all my requests 😛